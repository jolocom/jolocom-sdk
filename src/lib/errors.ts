import * as Sentry from '@sentry/react-native'
import VersionNumber from 'react-native-version-number'
import { sentryDSN } from 'src/config'
import { routeList } from '../routeList'
import strings from '../locales/strings'
import { ErrorCode } from './errors/index'
export { ErrorCode }

export class AppError extends Error {
  // private code: ErrorCode
  public origError: any
  public navigateTo: routeList
  public code: ErrorCode

  public constructor(
    code = ErrorCode.Unknown,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    origError?: any,
    navigateTo: routeList = routeList.AppInit,
  ) {
    super(strings[code] || strings[ErrorCode.Unknown])
    this.code = code
    this.origError = origError
    this.navigateTo = navigateTo
  }
}

export const errorTitleMessages = [strings.DAMN, strings.OH_NO, strings.UH_OH]

export function initErrorReporting() {
  Sentry.init({
    dsn: sentryDSN,
    release: `${VersionNumber.bundleIdentifier}@${VersionNumber.appVersion}`,

    // disable automatic reporting of errors/rejections without user consent
    integrations: defaultIntegrations =>
      defaultIntegrations.filter(i => i.name !== 'ReactNativeErrorHandlers'),

    /**
     * @param event
     * @dev TODO Decide if event.contexts.app should be stripped
     */

    beforeSend: event => {
      const { extra, contexts } = event

      if (extra && contexts) {
        if (!extra.sendPrivateData) {
          delete contexts.device
          delete contexts.os
        }

        delete extra.sendPrivateData
      }

      return { ...event, contexts, extra }
    },
  })
}

export interface UserReport {
  userError: string | undefined
  userDescription: string
  userContact: string
  sendPrivateData: boolean
}

interface ErrorReport extends UserReport {
  error: AppError | Error | undefined
}

export function reportError(report: ErrorReport) {
  Sentry.withScope(scope => {
    if (report.error instanceof AppError && report.error.origError) {
      scope.setExtras({
        AppError: report.error.message,
        UserError: report.userError,
        UserDescription: report.userDescription,
        UserContact: report.userContact,
        sendPrivateData: report.sendPrivateData,
      })
      report.error = report.error.origError
    }
    Sentry.captureException(report.error)
  })
}
