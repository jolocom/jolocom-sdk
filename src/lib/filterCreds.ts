import { buildTransform, Filter, Ordering } from './filter'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import { complement } from 'ramda'

const expiredFilter: Filter<SignedCredential> = cred =>
  cred.expires.valueOf() < new Date().valueOf()

const validFilter: Filter<SignedCredential> = complement(expiredFilter)

const issuerFilter = (issuerDid: string): Filter<SignedCredential> => cred =>
  cred.issuer === issuerDid

const typeFilter = (typ: string): Filter<SignedCredential> => cred =>
  cred.type.includes(typ)

const mostRecentOrder: Ordering<SignedCredential> = (c1, c2) =>
  c2.issued.valueOf() - c1.issued.valueOf()

/**
 * @dev Some basic predefined filters, if required they can be removed
 * and the filtering / ordering construction functions can be exposed instead
 */

export const filters = {
  filterByExpired: buildTransform([expiredFilter]),
  filterByValid: buildTransform([validFilter]),
  filterByIssuer: (did: string) => buildTransform([issuerFilter(did)]),
  filterByType: (typ: string) => buildTransform([typeFilter(typ)]),
  orderByRecent: buildTransform([mostRecentOrder]),
  documentFilter: (documentTypes: string[]) =>
    buildTransform(documentTypes.map(typeFilter)),
}
