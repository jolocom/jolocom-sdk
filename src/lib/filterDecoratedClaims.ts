import { buildTransform, Filter } from './filter'
import { DecoratedClaims } from './types'
import { complement } from 'ramda'

const expiredFilter: Filter<DecoratedClaims> = cred =>
  cred.expires ? cred.expires.valueOf() < new Date().valueOf() : true

const validFilter: Filter<DecoratedClaims> = complement(expiredFilter)

const issuerFilter = (issuerDid: string): Filter<DecoratedClaims> => cred =>
  cred.issuer.did === issuerDid

const typeFilter = (typ: string): Filter<DecoratedClaims> => cred =>
  cred.credentialType.includes(typ)

/**
 * @dev Some basic predefined filters, if required they can be removed
 * and the filtering / ordering construction functions can be exposed instead
 */

export const filters = {
  filterByExpired: buildTransform([expiredFilter]),
  filterByValid: buildTransform([validFilter]),
  filterByIssuer: (did: string) => buildTransform([issuerFilter(did)]),
  filterByType: (typ: string) => buildTransform([typeFilter(typ)]),
  documentFilter: (documentTypes: string[]) =>
    buildTransform(documentTypes.map(typeFilter)),
}
