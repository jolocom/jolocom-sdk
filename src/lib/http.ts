// @ts-ignore TODO
const fetch = global.fetch || require('node-fetch')

interface HttpAgent {
  getRequest<T>(endpoint: string): Promise<T>
  postRequest<T>(endpoint: string, headers: any, data: any): Promise<T>
  headRequest(endpoint: string): Promise<{ status: number }>
}

enum HttpMethods {
  GET = 'GET',
  POST = 'POST',
  HEAD = 'HEAD',
}

/**
 * @todo is this used anywhere?
 * @todo the res.json() prevents us from using these methods
 * Also
 */
export const httpAgent: HttpAgent = {
  getRequest: <T>(endpoint: string): Promise<T> =>
    fetch(endpoint, {
      method: HttpMethods.GET,
    }).then((res: { json: () => any }) => res.json()),
  headRequest(endpoint: string) {
    return fetch(endpoint, {
      method: HttpMethods.HEAD,
    }).then((res: { json: () => any }) => res.json())
  },
  postRequest<T>(endpoint: string, headers: any = {}, data: any): Promise<T> {
    const body = typeof data === 'string' ? data : JSON.stringify(data)
    return fetch(endpoint, {
      method: HttpMethods.POST,
      headers,
      body
    }).then(async (res: { ok: boolean, status: number, json: () => any, text: () => string }) => {
      if (!res.ok || res.status !== 200) {
        const err = new Error('http request failed')
        console.error('http request failed', err, res, '\nResponse Body', await res.text())
        throw err
      }
      return res.json()
    })
  },
}
