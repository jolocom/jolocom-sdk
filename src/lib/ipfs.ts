import { IIpfsConnector } from 'jolocom-lib/js/ipfs/types'
import { IDidDocumentAttrs } from 'jolocom-lib/js/identity/didDocument/types'

// @ts-ignore
const FormData = global.FormData || require('form-data')
// @ts-ignore
const fetch = global.fetch || require('node-fetch')

export class IpfsCustomConnector implements IIpfsConnector {
  private ipfsHost!: string

  constructor(config: { host: string; protocol: string; port: number }) {
    this.ipfsHost = `${config.protocol}://${config.host}:${config.port}`
  }

  async storeJSON({
    data,
    pin,
  }: {
    data: object
    pin: boolean
  }): Promise<string> {
    if (!data || typeof data !== 'object') {
      throw new Error(`JSON expected, received ${typeof data}`)
    }

    const endpoint = `${this.ipfsHost}/api/v0/add?pin=${pin}`

    // NOTE: we use FormData because it automatically creates the necessary
    // headers, which are not trivially
    // see https://github.com/github/fetch/issues/505#issuecomment-293064470
    const body = new FormData()
    body.append('ddo', Buffer.from(JSON.stringify(data)))

    const res = await fetch(endpoint, { method: 'POST', body })
    const resJson = await res.json()
    return resJson.Hash
  }

  async catJSON(hash: string): Promise<IDidDocumentAttrs> {
    const endpoint = `${this.ipfsHost}/api/v0/cat?arg=${hash}`
    const res = await fetch(endpoint, { method: 'GET' })
    return res.json()
  }

  async removePinnedHash(hash: string): Promise<void> {
    const resolutionGateway = 'https://ipfs.io/ipfs/'
    const endpoint = `${resolutionGateway}/${hash}`
    const res = await fetch(endpoint, { method: 'GET' })

    // @ts-ignore
    if (res.respInfo.status !== 200) {
      throw new Error(
        // @ts-ignore
        `Removing pinned hash ${hash} failed, status code: ${res.respInfo.status}`,
      )
    }
  }

  async createDagObject({
    data,
    pin,
  }: {
    data: object
    pin: boolean
  }): Promise<string> {
    if (!data || typeof data !== 'object') {
      throw new Error(`Object expected, received ${typeof data}`)
    }

    const endpoint = `${this.ipfsHost}/api/v0/dag/put?pin=${pin}`
    const body = new FormData()
    body.append('dag', Buffer.from(JSON.stringify(data)))

    const res = fetch(endpoint, { method: 'POST', body })
    const resJson = await res.json()
    return resJson.Cid['/']
  }

  async resolveIpldPath(pathToResolve: string): Promise<object> {
    const endpoint = `${this.ipfsHost}/api/v0/dag/get?arg=${pathToResolve}`
    const res = fetch(endpoint, { method: 'GET' })

    return await res.json()
  }
}
