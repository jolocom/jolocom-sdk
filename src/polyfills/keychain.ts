import { writeFileSync, readFileSync } from 'fs'
import path from 'path'

// from src/lib/keychain.ts
export class KeyChain implements KeyChainInterface {
  static PASSWORD_LOCATION = path.resolve(
    `${__dirname}/../local_identity.password.txt`,
  )
  private pass: string | null = null

  public async getPassword() {
    if (this.pass != null) return this.pass
    try {
      this.pass = readFileSync(KeyChain.PASSWORD_LOCATION).toString()
    } catch (err) {
      console.error('Error reading password file', err, '\n\n')
      throw err
    }

    return this.pass
  }

  public async savePassword(pass: string) {
    console.log('KEYCHAIN SAVED PASSWORD: ', pass)
    this.pass = pass
    writeFileSync(KeyChain.PASSWORD_LOCATION, pass)
  }
}

export interface KeyChainInterface {
  savePassword: (password: string) => Promise<void>
  getPassword: () => Promise<string>
}
