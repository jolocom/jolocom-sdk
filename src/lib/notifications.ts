import { AppError } from 'src/lib/errors'
import { randomBytes } from 'crypto'

interface NotificationMessage {
  title: string
  message?: string
}

/**
 * A notification is either dismissible or not
 * if it is dismissible it can also optionally have
 * - a label (for the dismiss button)
 * - a timeout (to auto-dismiss)
 * - a callback
 */

export interface NotificationDismiss {
  dismiss?:
    | false
    | {
        label?: string
        timeout?: number
        onDismiss?: (...args: any) => void
      }
}

/**
 * A notification may have a call to action button,
 * with a label and an onInteract callback
 * It is not possible to simply specify a boolean, because if an interaction is
 * expected then a callback and label are required.
 */
interface NotificationInteract {
  interact?: {
    label: string
    onInteract: (...args: any) => void | boolean | Promise<void | boolean>
  }
}

/**
 * The payload is either a message if the type is warning or info, or an error
 * if the type is error.
 */

type NotificationPayload = {
  type: NotificationType.warning | NotificationType.info
  error?: AppError
} & NotificationMessage

interface NotificationBase {
  id: string
  type: NotificationType
}

type PartialNotification = Partial<NotificationBase> &
  NotificationInteract &
  NotificationDismiss

const createNotificationFactory = (
  template: PartialNotification & Omit<NotificationBase, 'id'>,
) => (overrides: PartialNotification & NotificationMessage): Notification =>
  ({
    id: randomBytes(4).toString('hex'),
    ...(template as Notification),
    ...overrides,
  } as Notification)

export type Notification = NotificationBase &
  NotificationInteract &
  NotificationDismiss &
  NotificationPayload

export enum NotificationFilter {
  none,
  all,
  onlyDismissible,
}

export enum NotificationType {
  error = 'error',
  info = 'info',
  warning = 'warning',
}

export const createInfoNotification = createNotificationFactory({
  type: NotificationType.info,
  dismiss: {
    timeout: 3000,
  },
})

export const createWarningNotification = createNotificationFactory({
  type: NotificationType.warning,
  dismiss: {
    timeout: 3000,
  },
})

export const createStickyNotification = createNotificationFactory({
  type: NotificationType.info,
  dismiss: false,
})
