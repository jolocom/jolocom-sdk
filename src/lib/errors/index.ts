import { routeList } from '../../routeList'
import strings from '../../locales/strings'
import { ErrorReport, ErrorCode, IAppError } from './types'
export { ErrorCode }

import { reportErrorToSentry, initSentry } from '../../polyfills/sentry'

export class AppError extends Error implements IAppError {
  // private code: ErrorCode
  public origError: any
  public navigateTo: routeList

  public constructor(
    code = ErrorCode.Unknown,
    origError?: any,
    navigateTo: routeList = routeList.AppInit,
  ) {
    super(strings[code] || strings[ErrorCode.Unknown])
    // this.code = code
    this.origError = origError
    this.navigateTo = navigateTo
  }
}

export const errorTitleMessages = [strings.DAMN, strings.OH_NO, strings.UH_OH]

export function reportError(report: ErrorReport) {
  let extraData
  if (report.error instanceof AppError && report.error.origError) {
    extraData = {
      AppError: report.error.message,
      UserError: report.userError,
      UserDescription: report.userDescription,
      UserContact: report.userContact,
      sendPrivateData: report.sendPrivateData,
    }
    report.error = report.error.origError
  }
  return reportErrorToSentry(report, extraData)
}

export const initErrorReporting = initSentry
