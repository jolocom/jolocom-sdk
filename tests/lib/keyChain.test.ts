import { KeyChain } from 'src/lib/keychain'

describe('KeyChain lib', () => {
  let KC: KeyChain

  beforeEach(() => {
    KC = new KeyChain()
  })

  it('should contain the correct username', () => {
    // @ts-ignore private
    expect(KC.username).toBe('JolocomSmartWallet')
    // @ts-ignore private
    expect(KC.nativeLib).toBeDefined()
  })

  it('should correctly save the password', async () => {
    const mockSetGenericPassword = jest.fn()
    // @ts-ignore private
    KC.nativeLib.setGenericPassword = mockSetGenericPassword

    await KC.savePassword('test')

    expect(mockSetGenericPassword.mock.calls).toMatchSnapshot()
  })

  it('should correctly return if saving the password failed', async () => {
    // @ts-ignore private
    KC.nativeLib.setGenericPassword = async () => {
      throw new Error('scary error')
    }

    let error

    try {
      await KC.savePassword('test')
    } catch (err) {
      error = err
    }

    expect(error).toEqual(new Error('scary error'))
  })

  it('should correctly retrieve password', async () => {
    // @ts-ignore private
    KC.nativeLib.getGenericPassword = jest.fn().mockReturnValue({
      username: 'JolocomSmartWallet',
      password: 'test',
    })

    const result = await KC.getPassword()
    expect(result).toEqual('test')
  })

  it('should correctly return if password retrieval failed', async () => {
    // @ts-ignore private
    KC.nativeLib.getGenericPassword = jest.fn().mockReturnValue(false)

    let error

    try {
      await KC.getPassword()
    } catch (err) {
      error = err
    }
    expect(error).toEqual(
      new Error('Password could not be retrieved from the keychain'),
    )
  })
})
