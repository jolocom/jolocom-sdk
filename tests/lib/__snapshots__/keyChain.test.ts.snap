// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyChain lib should correctly save the password 1`] = `
Array [
  Array [
    "JolocomSmartWallet",
    "test",
  ],
]
`;
